packages/dds/task-manager/src/taskManager.ts

```diff
@@ -193,6 +203,24 @@ export class TaskManager extends SharedObject<ITaskManagerEvents> implements ITa
             this.removeClientFromQueue(taskId, clientId);
         });
 
+        this.opWatcher.on("completed", (taskId: string, clientId: string, local: boolean, messageId: number) => {
+            if (runtime.connected && local) {
+                const pendingOp = this.latestPendingOps.get(taskId);
+                assert(pendingOp !== undefined, "Unexpected op");
+                // TODO: check below comment and stuff, see if applicable
+                // Need to check the id, since it's possible to markAsComplete multiple times before the acks
+                if (messageId === pendingOp.messageId) {
+                    assert(pendingOp.type === "completed", "Unexpected op type");
+                    // Delete the pending, because we no longer have an outstanding op
+                    this.latestPendingOps.delete(taskId);
+                }
+            }
+
+            this.deleteQueueForTask(taskId);
+            this.completedWatcher.emit("completed", taskId);
+            this.emit("completed", taskId);
```

Comment:
The local client will be able to observe the change immediately via `queued()` and `haveTaskLock()`, so the "completed" event should probably be emitted immediately to reflect that (rather than waiting for the ack).

Reply:
Based on inspection I think this is probably already an issue with `abandon()` as well -- that it doesn't emit an observable event until getting the ack although the imperative APIs reveal that it's been abandoned.

The solution for both might be as simple as emitting a `this.queueWatcher.emit("queueChange"...` at the time they are invoked?  Might need some thought though.

Reply:
I think it may be simpler to have `completedWatcher`/`abandonWatcher` so we don't have to do any logic when listening to `queueWatcher` to know if the task was completed/abandoned. I'm not totally following the issue you outlined with the completed event not firing immediately, since it should fire pretty much right away.

Reply:
Yep you got it.  I think your change is on the right track but there might be some more state we need to track (see other comment)